File: backend/application/strategies/factory.py
from domain.interview_mode import InterviewMode
from application.strategies.base import InterviewStrategy
from application.strategies.resume_deep_dive import ResumeDeepDiveStrategy
from application.strategies.behavioral import BehavioralStrategy
from application.strategies.hr_screening import HRScreeningStrategy
from application.strategies.mixed import MixedInterviewStrategy

_STRATEGIES: dict[InterviewMode, type[InterviewStrategy]] = {
    InterviewMode.RESUME_DEEP_DIVE: ResumeDeepDiveStrategy,
    InterviewMode.BEHAVIORAL: BehavioralStrategy,
    InterviewMode.HR_SCREENING: HRScreeningStrategy,
    InterviewMode.MIXED: MixedInterviewStrategy,
}


class InterviewStrategyFactory:
    @staticmethod
    def get(mode: InterviewMode) -> InterviewStrategy:
        strategy_cls = _STRATEGIES.get(mode)
        if not strategy_cls:
            raise ValueError(f"Unsupported interview mode: {mode}")
        return strategy_cls()

File: backend-node/src/services/strategies/factory.ts
import { InterviewMode } from "../../models/interview-mode.js";
import type { InterviewStrategy } from "./base.js";
import { ResumeDeepDiveStrategy } from "./resume-deep-dive.js";
import { TechnicalStrategy } from "./technical.js";
import { HRScreeningStrategy } from "./hr-screening.js";
import { MixedInterviewStrategy } from "./mixed.js";

const STRATEGIES: Record<InterviewMode, () => InterviewStrategy> = {
  [InterviewMode.RESUME_DEEP_DIVE]: () => new ResumeDeepDiveStrategy(),
  [InterviewMode.TECHNICAL]: () => new TechnicalStrategy(),
  [InterviewMode.HR_SCREENING]: () => new HRScreeningStrategy(),
  [InterviewMode.MIXED]: () => new MixedInterviewStrategy(),
};

export class InterviewStrategyFactory {
  static get(mode: InterviewMode): InterviewStrategy {
    const factory = STRATEGIES[mode];
    if (!factory) {
      throw new Error(`Unsupported interview mode: ${mode}`);
    }
    return factory();
  }
}
